src/planner.cc

```diff
@@ -310,201 +409,146 @@ double Planner::Impl::euclidian_speed(const struct AxisTarget *t) {
-    target_pos->speed = peak_speed;  // Don't manage to accelerate to desired v
+  // We need to change the other axes steps for accel travel decel. We have to
+  // be careful of having the right rounding but at the same time accel + travel
+  // + decel has to stay the same. 1) find the deceleration ramp and the number
+  // of segments you can push in the backend.
+  uint32_t num_segments = planning_buffer_.size();
+  if (!flush_planning_queue) {
+    num_segments = 0;
   }
 
-  // Make sure the target feedrate for the move is clamped to what all the
-  // moving axes can reach.
-  const double max_speed = clamp_defining_axis_limit(
-    axis_steps, cfg_->max_feedrate, defining_axis, cfg_->steps_per_mm);
-  if (max_speed < target_pos->speed) target_pos->speed = max_speed;
-
-  const double accel_fraction =
-    (last_speed < target_pos->speed)
-      ? steps_for_speed_change(a, last_speed, &target_pos->speed,
-                               abs_defining_axis_steps) /
-          abs_defining_axis_steps
-      : 0;
-
-  // We only decelerate if the upcoming speed is _slower_
-  double dummy_next_speed = next_speed;  // Don't care to modify; we don't have
-  const double decel_fraction =
-    (next_speed < target_pos->speed)
-      ? steps_for_speed_change(-a, target_pos->speed, &dummy_next_speed,
-                               abs_defining_axis_steps) /
-          abs_defining_axis_steps
-      : 0;
-
 #if 0
-  // Useful debugging info.
-  // TODO: we get a decel glitch when the last_pos has accelerated to a speed
-  // that is faster than the target can decelerate from.
-  if (dummy_next_speed != next_speed) {
-    fprintf(stderr, "  \033[1m\033[31mGLITCH\033[0m "
-                    "- move too short for full decel (reached v1: %10.2f)\n",
-            dummy_next_speed);
+  fprintf(stderr, "\nTo be enqueued: %d/%zu.\n", num_segments, planning_buffer_.size());
+  for (unsigned i = 0; i < planning_buffer_.size(); ++i) {
+    fprintf(stderr, "%s\n", planning_buffer_[i]->ToJsonString().c_str());
   }
 #endif
 
-  assert(accel_fraction + decel_fraction <= 1.0 + 1e-3);
-
-#if 0
-  // fudging: if we have tiny acceleration segments, don't do these at all
-  // but only do speed; otherwise we have a lot of rattling due to many little
-  // segments of acceleration/deceleration (e.g. for G2/G3).
-  // This is not optimal. Ideally, we would actually calculate in terms of
-  // jerk and optimize to stay within that constraint.
-  const int accel_decel_steps
-    = (accel_fraction + decel_fraction) * abs_defining_axis_steps;
-  const double accel_decel_mm
-    = (accel_decel_steps / cfg_->steps_per_mm[defining_axis]);
-  const char do_accel = (accel_decel_mm > 2 || accel_decel_steps > 16);
-#else
-  const char do_accel = 1;
-#endif
-
-  bool has_accel = false;
-  bool has_move = false;
-  bool has_decel = false;
+  // We require a slot to be always present.
+  if (num_segments == 0 &&
+      (planning_buffer_.size() == PLANNING_BUFFER_CAPACITY - 1))
```

Comment:
With the latest version in head, we have now also planning_buffer_.capacity(), so this can be formulated nicer.

(also note, that capacity() of the buffer is actually `PLANNING_BUFFER_CAPACITY - 1`, as the RingDeque needs one empty slot internally)

Reply:
(solved)